crates/doc/src/inference.rs

```diff
@@ -1484,6 +1561,152 @@ impl Shape {
     }
 }
 
+const MAX_ROOT_FIELDS: usize = 750;
+const MAX_NESTED_FIELDS: usize = 200;
+
+impl Shape {
+    // Widen a Shape to make the provided AsNode value fit.
+    // Returns a hint if some locations might exceed their maximum allowable size.
+    // NOTE: If a particular location defines `additionalProperties` as a subschema, don't
+    // add the field from `AsNode`, instead jump straight to squashing the the field into `additionalProperties`
+    pub fn widen<'n, N>(&mut self, node: &'n N, loc: Location) -> bool
+    where
+        N: AsNode,
+    {
+        match node.as_node() {
+            crate::Node::Object(fields) => {
+                // See comment in `ObjShape::widen` about when to set `is_required`
+                // on newly encountered fields. Detects whether this is the
+                // very first time this location has seen an OBJECT shape.
+                let is_first_time = !self.type_.overlaps(types::OBJECT);
+                self.type_ = self.type_ | types::OBJECT;
+
+                if let None = self.object.additional {
+                    self.object.additional = Some(Box::new(Shape {
+                        type_: types::INVALID,
+                        provenance: Provenance::Inline,
+                        ..Default::default()
+                    }))
+                }
+
+                let hint = self.object.widen::<N>(fields, loc, is_first_time);
+
+                match (hint, loc) {
```

Comment:
nit: does this belong in ObjShape::widen ?

Reply:
Yeah, I think it does